{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Running questions\n",
    "### How to handle Ns? [We're probably going to ignore them]\n",
    "### Should I remove reference from alternate the same way I remove alternate from the reference? \n",
    "### Find longest string rather than checking entire stretch and neighbors to see if they're homopolymers? \n",
    "\n",
    "### Note: I will clean up stuff for sure (moving lines to functions, etc), this is more about logic "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cyvcf2\n",
    "from pyfaidx import Fasta\n",
    "import os\n",
    "import argparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['20_shared_rare_1SCS014.vcf',\n",
       " '1SCS014.indels_removed_homopolymer.vcf',\n",
       " 'YL99029.indels.vcf',\n",
       " 'GRCh38_simplerepeats_1-6_homopolymers_1bpslop.bed',\n",
       " 'XH9R013.indels.vcf',\n",
       " '1SCS014.reliable_indels.vcf',\n",
       " 'sites_in_more_than_5_cut.bed',\n",
       " 'GRCh38_simplerepeats_1-6_homopolymers.bed',\n",
       " '.git',\n",
       " 'XH9R013.reliable_indels_any_overlap.vcf',\n",
       " 'UA6U012.reliable_indels.vcf',\n",
       " 'XH9R013.reliable_indels.vcf',\n",
       " 'GRCh38_simplerepeats_1-6_homopolymers_5bpslop.bed',\n",
       " '1SCS014.indels.vcf',\n",
       " '1SCS014.bedtools.save',\n",
       " 'K8J7003.indels.vcf',\n",
       " 'GRCh38_simplerepeats_1-6_homopolymers_2bpslop.bed',\n",
       " 'K8J7003.reliable_indels_any_overlap.vcf',\n",
       " 'K8J7003.reliable_indels.vcf',\n",
       " 'YL99029.reliable_indels.vcf',\n",
       " 'YL99029.reliable_indels_any_overlap.vcf',\n",
       " '1SCS014.trustworthyindels.vcf',\n",
       " 'UA6U012.reliable_indels_any_overlap.vcf',\n",
       " 'sites_in_more_than_5.bed',\n",
       " 'shared_blood_sperm_mutation_analysis.bed',\n",
       " '1SCS014.reliable_indels_any_overlap.vcf',\n",
       " 'UA6U012.indels.vcf']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.chdir('/Users/quinlan/u1264408/Git/spermseq')\n",
    "os.listdir()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "A = '1SCS014.indels.vcf'\n",
    "B = 'K8J7003.indels.vcf'\n",
    "C = 'UA6U012.indels.vcf'\n",
    "D = 'YL99029.indels.vcf'\n",
    "E = 'XH9R013.indels.vcf'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = Fasta('/Users/quinlan/u1264408/human_g1k_v38_decoy_phix.fasta', as_raw=True) # look into as_RAW\n",
    "# as_raw: optional parameter to specify whether to return sequences as a Sequence() object or as a raw string.\n",
    "from cyvcf2 import VCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig 'chr1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] FILTER '-' is not defined in the header\n",
      "[W::vcf_parse] INFO 'donor' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse] INFO 'variation_type' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse] INFO 'subtype' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse_format] FORMAT 'filter' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse_format] FORMAT 'sample' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse_format] FORMAT 'AF' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse_format] FORMAT 'alt_depth' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse_format] FORMAT 'depth' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse_format] FORMAT 'no_calls' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse_format] FORMAT 'tissue' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse_format] FORMAT 'context' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse_format] FORMAT 'qual' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse_format] FORMAT 'age' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse] Contig 'chr2' is not defined in the header. (Quick workaround: index the file with tabix.)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FAIL  chr1 75806922 T TA\n",
      "TTAA\n",
      "PASS chr1 75881235 C CAT\n",
      "AACGTG\n",
      "FAIL  chr1 100353868 G GT\n",
      "TGTT\n",
      "FAIL  chr1 243572906 TA T\n",
      "TA\n",
      "FAIL  chr1 243572907 AC A\n",
      "AC\n",
      "FAIL  chr1 243572908 C CT\n",
      "ACTT\n",
      "FAIL  chr1 243572908 CT C\n",
      "CT\n",
      "FAIL  chr1 243572908 CTT C\n",
      "CT\n",
      "FAIL  chr2 38997474 G GT\n",
      "AGTT\n",
      "FAIL  chr2 39007201 T TA\n",
      "ATAA\n",
      "FAIL  chr2 39013443 TAA T\n",
      "TA\n",
      "FAIL  chr2 39013443 TA T\n",
      "TA\n",
      "FAIL  chr2 39013443 T TA\n",
      "CTAA\n",
      "FAIL  chr2 39035155 CA C\n",
      "CA\n",
      "FAIL  chr2 39035155 C CA\n",
      "GCAA\n",
      "FAIL  chr2 39035331 AT A\n",
      "AT\n",
      "FAIL  chr2 47408384 C CT\n",
      "ACTT\n",
      "FAIL  chr2 47408384 CT C\n",
      "CT\n",
      "FAIL  chr2 47408384 CTT C\n",
      "CT\n",
      "FAIL  chr2 47414420 TAAA T\n",
      "TA\n",
      "FAIL  chr2 47414420 TAA T\n",
      "TA\n",
      "FAIL  chr2 47474913 T TAGG\n",
      "AATTAGGA\n",
      "FAIL  chr2 47475310 TTG T\n",
      "TT\n",
      "FAIL  chr2 47475311 TG T\n",
      "TG\n",
      "FAIL  chr2 47475312 G GT\n",
      "TGTT\n",
      "FAIL  chr2 47475312 GT G\n",
      "GT\n",
      "FAIL  chr2 47791154 TTG T\n",
      "TT\n",
      "FAIL  chr2 47791154 T TTG\n",
      "GTTTGT\n",
      "FAIL  chr2 47804889 CT C\n",
      "CT\n",
      "FAIL  chr2 47805601 A AT\n",
      "GATT\n",
      "FAIL  chr2 47805601 A ATT\n",
      "TGATTT\n",
      "FAIL  chr2 47805601 AT A\n",
      "AT\n",
      "FAIL  chr2 47805601 ATT A\n",
      "AT\n",
      "FAIL  chr2 47805734 GACTA G\n",
      "GA\n",
      "FAIL  chr2 47805735 ACTAT A\n",
      "AC\n",
      "FAIL  chr2 47806407 C CT\n",
      "ACTT\n",
      "FAIL  chr2 189805813 TAA T\n",
      "TA\n",
      "FAIL  chr2 189805813 TA T\n",
      "TA\n",
      "FAIL  chr2 189805813 T TA\n",
      "TTAA\n",
      "FAIL  chr2 189852791 GA G\n",
      "GA\n",
      "FAIL  chr2 189852791 G GA\n",
      "AGAA\n",
      "FAIL  chr2 189863889 TA T\n",
      "TA\n",
      "FAIL  chr3 12609348 TA T\n",
      "TA\n",
      "FAIL  chr3 37020383 G GAGAGCATCCTGGAGCGGGTGCAGCAGCACATCGAGAGCAAGCTCCTGGGCTCCAA\n",
      "GAATGTGGATGTTAATGTGCACCCCACAAAGCATGAAGTTCACTTCCTGCACGAGGAGAGCATCCTGGAGCGGGTGCAGCAGCACATCGAGAGCAAGCTCCTGGGCTCCAAT\n",
      "FAIL  chr3 41227185 GAT G\n",
      "GA\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig 'chr3' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr4' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr5' is not defined in the header. (Quick workaround: index the file with tabix.)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FAIL  chr3 50118454 TGAGA T\n",
      "TG\n",
      "FAIL  chr3 50118454 TGA T\n",
      "TG\n",
      "FAIL  chr3 50118454 T TGA\n",
      "AGTGAG\n",
      "FAIL  chr4 1804285 TG T\n",
      "TG\n",
      "FAIL  chr4 1804285 T TG\n",
      "GTGG\n",
      "FAIL  chr4 54731256 T TAG\n",
      "GGTAGA\n",
      "FAIL  chr4 54732045 A AT\n",
      "GATT\n",
      "FAIL  chr5 80675095 C CA\n",
      "ACAA\n",
      "FAIL  chr5 80864776 AAAAT A\n",
      "AA\n",
      "FAIL  chr5 80864777 AAATG A\n",
      "AA\n",
      "FAIL  chr5 112775612 TA T\n",
      "TA\n",
      "FAIL  chr5 112775612 T TA\n",
      "TTAA\n",
      "FAIL  chr5 112839514 TAAAAG T\n",
      "TA\n",
      "FAIL  chr5 112840921 TA T\n",
      "TA\n",
      "FAIL  chr5 112841744 GA G\n",
      "GA\n",
      "FAIL  chr5 112842691 A ATACATCTCCAGGTAGACAGATGAGCCAACAGAACCTTACCAAACAAACAGGTTTATCCAAGAATGCCAGTAGTATTCCAAGAAGTGAGTCTGCCTCCAAAGGACTAAATCAGATGAATAATGGTAATGGAGCCAATAAAAAGGTAGAACTTTCTAGAATGTCTTCAACTAAATCAAGTGGAAGTGAATCTGATAGATCAGAAAGACCTGTATTAGTACGCCAGTCAACTTTCATCAAAGAAGCTCCAAGCCCAACCTTAAGAAGAAAATTGGAGGAATCTGCTTCATTTGAATCTCTTTCTCCATCATCTAGACCAGCTTCTCCCACTAGGTCCCAGGCACAAACTCCAGTTTTAAG\n",
      "CTAAAAAAGGCCCACCCCTTAAGACTCCAGCCTCCAAAAGCCCTAGTGAAGGTCAAACAGCCACCACTTCTCCTAGAGGAGCCAAGCCATCTGTGAAATCAGAATTAAGCCCTGTTGCCAGGCAGACATCCCAAATAGGTGGGTCAAGTAAAGCACCTTCTAGATCAGGATCTAGAGATTCGACCCCTTCAAGACCTGCCCAGCAACCATTAAGTAGACCTATACAGTCTCCTGGCCGAAACTCAATTTCCCCTGGTAGAAATGGAATAAGTCCTCCTAACAAATTATCTCAACTTCCAAGGACATCATCCCCTAGTACTGCTTCAACTAAGTCCTCAGGTTCTGGAAAAATGTCATATACATCTCCAGGTAGACAGATGAGCCAACAGAACCTTACCAAACAAACAGGTTTATCCAAGAATGCCAGTAGTATTCCAAGAAGTGAGTCTGCCTCCAAAGGACTAAATCAGATGAATAATGGTAATGGAGCCAATAAAAAGGTAGAACTTTCTAGAATGTCTTCAACTAAATCAAGTGGAAGTGAATCTGATAGATCAGAAAGACCTGTATTAGTACGCCAGTCAACTTTCATCAAAGAAGCTCCAAGCCCAACCTTAAGAAGAAAATTGGAGGAATCTGCTTCATTTGAATCTCTTTCTCCATCATCTAGACCAGCTTCTCCCACTAGGTCCCAGGCACAAACTCCAGTTTTAAGT\n",
      "FAIL  chr6 31741335 TTACA T\n",
      "TT\n",
      "FAIL  chr6 31741336 TACACACACACACACACAC T\n",
      "TA\n",
      "FAIL  chr6 31741336 TACACACACACACAC T\n",
      "TA\n",
      "FAIL  chr6 31741336 TACACACACACAC T\n",
      "TA\n",
      "FAIL  chr6 31741336 TACACACACAC T\n",
      "TA\n",
      "FAIL  chr6 31741336 TACACACAC T\n",
      "TA\n",
      "FAIL  chr6 31741336 TACACAC T\n",
      "TA\n",
      "FAIL  chr6 31741336 TACAC T\n",
      "TA\n",
      "FAIL  chr6 151808193 TG T\n",
      "TG\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig 'chr6' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr7' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr8' is not defined in the header. (Quick workaround: index the file with tabix.)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FAIL  chr6 151842588 TC T\n",
      "TC\n",
      "FAIL  chr6 152060952 C CT\n",
      "TCTT\n",
      "FAIL  chr6 159816629 AC A\n",
      "AC\n",
      "FAIL  chr6 159816629 ACT A\n",
      "AC\n",
      "FAIL  chr6 159816630 CT C\n",
      "CT\n",
      "FAIL  chr6 159816630 CTT C\n",
      "CT\n",
      "FAIL  chr6 159816630 CTTT C\n",
      "CT\n",
      "FAIL  chr6 159816630 CTTTT C\n",
      "CT\n",
      "FAIL  chr7 5992106 GC G\n",
      "GC\n",
      "FAIL  chr7 5997264 TA T\n",
      "TA\n",
      "FAIL  chr7 5997425 TG T\n",
      "TG\n",
      "FAIL  chr7 5997426 GAA G\n",
      "GA\n",
      "FAIL  chr7 5997426 GA G\n",
      "GA\n",
      "FAIL  chr7 5997426 G GA\n",
      "TGAA\n",
      "FAIL  chr7 140734774 CA C\n",
      "CA\n",
      "FAIL  chr7 140734774 C CA\n",
      "ACAA\n",
      "FAIL  chr7 140734797 G GA\n",
      "AGAA\n",
      "FAIL  chr7 140783126 AG A\n",
      "AG\n",
      "FAIL  chr8 22318323 A AT\n",
      "TATT\n",
      "FAIL  chr8 22318323 AT A\n",
      "AT\n",
      "FAIL  chr8 30847884 TATA T\n",
      "TA\n",
      "FAIL  chr8 30860058 CT C\n",
      "CT\n",
      "FAIL  chr8 30860059 TA T\n",
      "TA\n",
      "FAIL  chr8 30860059 T TA\n",
      "CTAA\n",
      "FAIL  chr9 842228 TA T\n",
      "TA\n",
      "FAIL  chr9 842229 AG A\n",
      "AG\n",
      "FAIL  chr9 842230 G GT\n",
      "AGTT\n",
      "FAIL  chr9 842230 GT G\n",
      "GT\n",
      "FAIL  chr9 842230 GTT G\n",
      "GT\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig 'chr9' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr10' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr11' is not defined in the header. (Quick workaround: index the file with tabix.)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FAIL  chr9 83973270 A AT\n",
      "AATT\n",
      "FAIL  chr9 83973270 AT A\n",
      "AT\n",
      "FAIL  chr9 83973270 ATT A\n",
      "AT\n",
      "FAIL  chr9 124482702 G GC\n",
      "CGCC\n",
      "FAIL  chr10 43100386 TG T\n",
      "TG\n",
      "FAIL  chr10 43100388 A AT\n",
      "GATT\n",
      "FAIL  chr10 43100388 A ATT\n",
      "TGATTT\n",
      "FAIL  chr10 43100388 AT A\n",
      "AT\n",
      "FAIL  chr10 43100388 ATT A\n",
      "AT\n",
      "FAIL  chr10 121551219 C CA\n",
      "GCAA\n",
      "FAIL  chr11 534403 GCCCAGGCCCAGC G\n",
      "GC\n",
      "FAIL  chr11 32430432 CAG C\n",
      "CA\n",
      "FAIL  chr11 32430451 GGA G\n",
      "GG\n",
      "FAIL  chr11 32430451 GGAGGGA G\n",
      "GG\n",
      "FAIL  chr11 32430453 AGG A\n",
      "AG\n",
      "FAIL  chr11 32430455 GGAGAGAGA G\n",
      "GG\n",
      "FAIL  chr11 32430455 GGAGAGA G\n",
      "GG\n",
      "FAIL  chr11 32430455 GGAGA G\n",
      "GG\n",
      "FAIL  chr11 32430455 GGA G\n",
      "GG\n",
      "FAIL  chr11 94479764 T TA\n",
      "TTAA\n",
      "FAIL  chr11 94486159 T TA\n",
      "ATAA\n",
      "FAIL  chr11 94616473 AT A\n",
      "AT\n",
      "FAIL  chr11 108227582 A AT\n",
      "TATT\n",
      "FAIL  chr11 108227582 AT A\n",
      "AT\n",
      "FAIL  chr11 108227732 TAA T\n",
      "TA\n",
      "FAIL  chr11 108229378 G GT\n",
      "AGTT\n",
      "FAIL  chr11 108317539 A AT\n",
      "GATT\n",
      "FAIL  chr11 108317539 AT A\n",
      "AT\n",
      "FAIL  chr11 108325249 GT G\n",
      "GT\n",
      "FAIL  chr11 108325983 TTCA T\n",
      "TT\n",
      "FAIL  chr11 108325985 CATT C\n",
      "CA\n",
      "FAIL  chr11 108326230 G GT\n",
      "TGTT\n",
      "FAIL  chr11 116835717 AC A\n",
      "AC\n",
      "FAIL  chr11 119206522 GCAC G\n",
      "GC\n",
      "FAIL  chr11 119274079 AG A\n",
      "AG\n",
      "FAIL  chr11 119274079 AGC A\n",
      "AG\n",
      "FAIL  chr11 119274080 GC G\n",
      "GC\n",
      "FAIL  chr11 119274081 C CT\n",
      "GCTT\n",
      "FAIL  chr11 119274081 CT C\n",
      "CT\n",
      "FAIL  chr11 119274753 G GT\n",
      "TGTT\n",
      "FAIL  chr11 119274753 GT G\n",
      "GT\n",
      "FAIL  chr11 119278645 TATG T\n",
      "TA\n",
      "FAIL  chr12 54282381 G GT\n",
      "TGTT\n",
      "FAIL  chr12 54282381 GT G\n",
      "GT\n",
      "FAIL  chr12 101729122 AT A\n",
      "AT\n",
      "FAIL  chr12 101734904 T TA\n",
      "CTAA\n",
      "FAIL  chr12 130345740 AT A\n",
      "AT\n",
      "FAIL  chr12 130345784 AT A\n",
      "AT\n",
      "FAIL  chr12 130346594 GA G\n",
      "GA\n",
      "FAIL  chr12 130354608 TG T\n",
      "TG\n",
      "PASS chr12 130362723 A AC\n",
      "TAGA\n",
      "FAIL  chr12 132657803 CAT C\n",
      "CA\n",
      "FAIL  chr12 132659279 TG T\n",
      "TG\n",
      "FAIL  chr12 132661165 CT C\n",
      "CT\n",
      "FAIL  chr12 132661166 TG T\n",
      "TG\n",
      "FAIL  chr12 132661167 GAAA G\n",
      "GA\n",
      "FAIL  chr12 132661167 GAA G\n",
      "GA\n",
      "FAIL  chr12 132661167 GA G\n",
      "GA\n",
      "FAIL  chr12 132661167 G GA\n",
      "TGAA\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig 'chr12' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr13' is not defined in the header. (Quick workaround: index the file with tabix.)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FAIL  chr12 132664195 CACCCAGTGTGTGGCCTCCAGCCTTCCCTCCTTCCTTCCT C\n",
      "CA\n",
      "FAIL  chr13 32339699 CA C\n",
      "CA\n",
      "FAIL  chr13 32341273 CAATT C\n",
      "CA\n",
      "FAIL  chr13 32356498 C CGTCTTTCCACAGCCAGGCAGTCTGTATCTTGCAAAAACATCCACTCTGCCTCGAATCTCTCTGAAAGCAGCAGTAGGAGGCCAAGTTCCCTCTGCGTGTTCTCATAAACAGGTATGTGTTTGTCTACAATACTGATGGCTTTTATGACAGAGTGTAATTTTATTTCATTAACTAGTATCTACAAATGGCTTTGTTTAAAGAATGAACACATTAGTGCAGGAATGGATGAATGAAATCATCATATTTTCTAATTAGCCTGCAGTGGCAGCCTCTGGCCCCTTGCTAGGCCTGCCTCATCCTGCTAAAGTGATCTGTGCTTCCAAATTACTACTTCTTTTCCCCCTTCAAATCTTTCTTATTTTGTCATTGTAAATGCTCTCAGCTAGGTGTTAAAGTAGTCTTACTGATATTCAAATGTGAATAACTGATAGCCCTGAACCTTCTATGAGCTATTTATATTTTCCAAAGAGGATTCTCCTTAAGCCAATATTATCTAGGTAGAATTTTAGGCAATGGAGAGGTGAAAATAATATTGATGACATTAATAGCTAACTTTGAGCATTTTCTAG\n",
      "CAAGTTGAAACATCCTTTCATTTATGAAGAATTAAAAGGGGTACCCTTTTTAGAGAAAAGGAGAGCATGTAAACTTCGAGGAAATTGATATGTATAATTTTATAAAACAGGGCTTGCGCTTTTTTTTTTTTGAGACAGAGTTTCGCTCTTGTTGCCCAGGCTGGAGTGCAATGGTGCAACCTCGGCTCACCGCAACCTCCTCCTCCCGAGTTCAAGTGATTCTCCTGCCTCAGCCTGCTGAATAGCTGGGATTACAGGCATGTGCCACCACACCTGGCTACTTTTGTGTTTTTTTTACTTTTATATATTTTTTTTTTGTTTAGTAGAGACAGGGTTTCTCCATTTTGGTCAGGCTGGTCTTGAACTCCCGACCTCAGATGATCTGCCCGCCTCAGCCTCCCAAAGTGCTGGGATTACAGGCGTGAGCCACTGTGCCTGGCCAGGGGTTGTGCTTTTTAAATTTCAATTTTATTTTTGCTAAGTATTTATTCTTTGATAGATTTAATTACAAGTCTTCAGAATGCCAGAGATATACAGGATATGCGAATTAAGAAGAAACAAAGGCAACGCGTCTTTCCACAGCCAGGCAGTCTGTATCTTGCAAAAACATCCACTCTGCCTCGAATCTCTCTGAAAGCAGCAGTAGGAGGCCAAGTTCCCTCTGCGTGTTCTCATAAACAGGTATGTGTTTGTCTACAATACTGATGGCTTTTATGACAGAGTGTAATTTTATTTCATTAACTAGTATCTACAAATGGCTTTGTTTAAAGAATGAACACATTAGTGCAGGAATGGATGAATGAAATCATCATATTTTCTAATTAGCCTGCAGTGGCAGCCTCTGGCCCCTTGCTAGGCCTGCCTCATCCTGCTAAAGTGATCTGTGCTTCCAAATTACTACTTCTTTTCCCCCTTCAAATCTTTCTTATTTTGTCATTGTAAATGCTCTCAGCTAGGTGTTAAAGTAGTCTTACTGATATTCAAATGTGAATAACTGATAGCCCTGAACCTTCTATGAGCTATTTATATTTTCCAAAGAGGATTCTCCTTAAGCCAATATTATCTAGGTAGAATTTTAGGCAATGGAGAGGTGAAAATAATATTGATGACATTAATAGCTAACTTTGAGCATTTTCTAGG\n",
      "FAIL  chr13 48379519 CA C\n",
      "CA\n",
      "FAIL  chr13 48379519 C CAA\n",
      "CTCAAA\n",
      "FAIL  chr13 48379519 C CA\n",
      "TCAA\n",
      "FAIL  chr13 48380023 CT C\n",
      "CT\n",
      "FAIL  chr13 48380146 CT C\n",
      "CT\n",
      "FAIL  chr13 48381227 C CT\n",
      "ACTT\n",
      "PASS chr13 48381523 C CAA\n",
      "AACATC\n",
      "FAIL  chr13 48459886 A AT\n",
      "GATT\n",
      "PASS chr13 48459886 A ATTTCT\n",
      "AACTGATTCTTT\n",
      "FAIL  chr13 48459886 ATTC A\n",
      "AT\n",
      "FAIL  chr14 45159080 TTA T\n",
      "TT\n",
      "FAIL  chr14 75018965 T TTAAACTTAATGGCCCCTAAATGAAAGACAGAAACAAGAAGGTTATAGTGTA\n",
      "GAGGACAGAGCTTCAATAAGGCGGCAACTTTCCTGTAAGCTCAGGCCATCATTAAACTTAATGGCCCCTAAATGAAAGACAGAAACAAGAAGGTTATAGTGTAT\n",
      "FAIL  chr14 75030725 TA T\n",
      "TA\n",
      "FAIL  chr14 75030725 T TA\n",
      "TTAA\n",
      "FAIL  chr14 75047900 C CT\n",
      "TCTT\n",
      "FAIL  chr14 81108343 T TTC\n",
      "CATTCT\n",
      "FAIL  chr14 81108343 T TTCTC\n",
      "TTCATTCTCT\n",
      "FAIL  chr14 81108494 C CT\n",
      "TCTT\n",
      "FAIL  chr14 81108494 CT C\n",
      "CT\n",
      "PASS chr14 81108501 T TTTTTTCTTTTTAC\n",
      "TTTTTTCTTTTTTTTTTTTTGGAATAAA\n",
      "PASS chr14 81108501 T TTTTTTTCTTTTTAC\n",
      "TTTTTTTCTTTTTTTTTTTTTGGAATAAAT\n",
      "FAIL  chr14 95099730 TTACA T\n",
      "TT\n",
      "FAIL  chr14 95099731 TACACACACAC T\n",
      "TA\n",
      "FAIL  chr14 95099731 TACACAC T\n",
      "TA\n",
      "FAIL  chr14 95099731 TACAC T\n",
      "TA\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig 'chr14' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr15' is not defined in the header. (Quick workaround: index the file with tabix.)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FAIL  chr14 95105297 AG A\n",
      "AG\n",
      "FAIL  chr14 95105298 GA G\n",
      "GA\n",
      "FAIL  chr14 95105298 G GA\n",
      "AGAA\n",
      "FAIL  chr14 95132538 CT C\n",
      "CT\n",
      "FAIL  chr14 103966662 CT C\n",
      "CT\n",
      "FAIL  chr14 103998790 GC G\n",
      "GC\n",
      "FAIL  chr14 103998791 C CT\n",
      "GCTT\n",
      "FAIL  chr14 103998791 CT C\n",
      "CT\n",
      "FAIL  chr14 103998791 CTT C\n",
      "CT\n",
      "FAIL  chr14 104004361 GTTTA G\n",
      "GT\n",
      "FAIL  chr14 104004361 GTTTATTTA G\n",
      "GT\n",
      "FAIL  chr15 89273310 TA T\n",
      "TA\n",
      "FAIL  chr15 89273511 TAAAA T\n",
      "TA\n",
      "FAIL  chr15 89273511 TAAA T\n",
      "TA\n",
      "FAIL  chr15 89273511 TAA T\n",
      "TA\n",
      "FAIL  chr15 89273511 TA T\n",
      "TA\n",
      "FAIL  chr15 89307731 A AG\n",
      "CAGG\n",
      "FAIL  chr16 89792607 GT G\n",
      "GT\n",
      "FAIL  chr16 89795865 CA C\n",
      "CA\n",
      "FAIL  chr17 41847791 C CACTGCACCCGGCTGAAAAGCTTTTTTTCATGGAAGATAATGGAAGAGGGAGTTATGAGATCACTGGTACCCCCAACAAGGGCTTCCTCAGTGAACAGAATCAATGGTTAGCAGTCAACCGTGTTTCTTTTGCTATCCACAGGTAGGTGGCTTTGATGGAGCTAATCGACTTAGGAGTGCCGAAGCCTACAGCCCTGTGGCTAACACTTGGCGCACAATCCCCACTATGTTTAATCCTCGTAGCAATTTTGGCATCGAGGTGGTGGATGACCTCTTGTTTGTGGTGGGTGGCTTTAATGGTTTTACCACCACCTTTAATGTTGAGTGCTATGATGAAAAGACCGATGAGTGGTATGATGCTCATGACATGAGTATATACCGCAGTGCTCTGAGCTGCTGTGTAGT\n",
      "GCTTATGGAGAACATGTATATGCGGTAAGTTTATCTAGTACCACACACACACAAAAAACACATTACCCCTAGATTTTGTATTAGTAAATGGGTTTATGCTACTATTGGTAAGTATTTGAAAAGCTTTTTTTTTTTTTGAGATGGAGTCTCGCTCTGTTGCCCAGGCTGTACTGCAGTGGCAGGATCTCGGCTCACTGCAAGCTCTGCCTCCCGGGTTCATGCCATTCTCCTGCCTCAGCCTCCCGAGCAGCTGGGACTACCGGCACCCGCCACCATGCCCAGCTAATTTTTTGTATTTTTAGTAGAGACAGGGTTTCACTGTGTTAGCCCAGATGGTCTCGATCTCCTGACTTTGTGATCCACCCGCCTCGGCCTCCCAGAGTGCTGGGATTACAGGCGTGAGCCACTGCACCCGGCTGAAAAGCTTTTTTTCATGGAAGATAATGGAAGAGGGAGTTATGAGATCACTGGTACCCCCAACAAGGGCTTCCTCAGTGAACAGAATCAATGGTTAGCAGTCAACCGTGTTTCTTTTGCTATCCACAGGTAGGTGGCTTTGATGGAGCTAATCGACTTAGGAGTGCCGAAGCCTACAGCCCTGTGGCTAACACTTGGCGCACAATCCCCACTATGTTTAATCCTCGTAGCAATTTTGGCATCGAGGTGGTGGATGACCTCTTGTTTGTGGTGGGTGGCTTTAATGGTTTTACCACCACCTTTAATGTTGAGTGCTATGATGAAAAGACCGATGAGTGGTATGATGCTCATGACATGAGTATATACCGCAGTGCTCTGAGCTGCTGTGTAGTA\n",
      "FAIL  chr17 42337874 AACTCCTTGACCTGAGGGAAT A\n",
      "AA\n",
      "FAIL  chr17 42339255 CAA C\n",
      "CA\n",
      "FAIL  chr17 42339255 CA C\n",
      "CA\n",
      "FAIL  chr17 42339255 C CA\n",
      "TCAA\n",
      "FAIL  chr17 43090921 GCA G\n",
      "GC\n",
      "FAIL  chr17 43093569 CT C\n",
      "CT\n",
      "FAIL  chr17 43104057 CA C\n",
      "CA\n",
      "FAIL  chr17 43104057 C CA\n",
      "TCAA\n",
      "FAIL  chr17 43104058 AAAAAAAAAAAAG A\n",
      "AA\n",
      "PASS chr17 43104059 AAAAAAAAAAAGAAAAAAAAAAGAAAAG NAAAAAAAAANNA\n",
      "GACTCCATCTCAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104060 AAAAAAAAAAGAAAAAAAAAAGAAAAG NAAAAAAAAGAA\n",
      "CTCCATCTCAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104060 AAAAAAAAAAGAAAAAAAAAAGAAAAG NAAAAAAAAGNA\n",
      "CTCCATCTCAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104060 AAAAAAAAAAGAAAAAAAAAAGAAAAG NAAAAAAANGAA\n",
      "CTCCATCTCAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104060 AAAAAAAAAAGAAAAAAAAAAGAAAAG NAAAAAANNGAA\n",
      "CTCCATCTCAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104060 AAAAAAAAAAGAAAAAAAAAAGAAAAG NAAAANAANNAA\n",
      "CTCCATCTCAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104060 AAAAAAAAAAGAAAAAAAAAAGAAAAG NAAANNAANGAA\n",
      "CTCCATCTCAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104060 AAAAAAAAAAGAAAAAAAAAAGAAAAG NAANANAAANNA\n",
      "CTCCATCTCAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104060 AAAAAAAAAAGAAAAAAAAAAGAAAAG NANAAAANAGAA\n",
      "CTCCATCTCAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104060 AAAAAAAAAAGAAAAAAAAAAGAAAAG NNAANAAANGNA\n",
      "CTCCATCTCAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104061 AAAAAAAAAGAAAAAAAAAAGAAAAG NAAANAAAGAA\n",
      "CCATCTCAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104062 AAAAAAAAGAAAAAAAAAAGAAAAG NAAAAAAGAA\n",
      "ATCTCAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104062 AAAAAAAAGAAAAAAAAAAGAAAAG NAAAAAANAA\n",
      "ATCTCAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104062 AAAAAAAAGAAAAAAAAAAGAAAAG NAAAAAANNA\n",
      "ATCTCAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104062 AAAAAAAAGAAAAAAAAAAGAAAAG NAAAAANGAA\n",
      "ATCTCAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104062 AAAAAAAAGAAAAAAAAAAGAAAAG NAAANNNGAA\n",
      "ATCTCAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104062 AAAAAAAAGAAAAAAAAAAGAAAAG NAANAANGAA\n",
      "ATCTCAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104063 AAAAAAAGAAAAAAAAAAGAAAAG NAAAAAGAA\n",
      "CTCAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104063 AAAAAAAGAAAAAAAAAAGAAAAG NAAAAAGNA\n",
      "CTCAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104063 AAAAAAAGAAAAAAAAAAGAAAAG NAAAANGAA\n",
      "CTCAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104063 AAAAAAAGAAAAAAAAAAGAAAAG NAAANNGAA\n",
      "CTCAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104063 AAAAAAAGAAAAAAAAAAGAAAAG NANANAGNA\n",
      "CTCAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104063 AAAAAAAGAAAAAAAAAAGAAAAG NNAAAANAN\n",
      "CTCAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104064 AAAAAAGAAAAAAAAAAGAAAAG NAAAAGNA\n",
      "CAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104064 AAAAAAGAAAAAAAAAAGAAAAG NAAAANAA\n",
      "CAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104064 AAAAAAGAAAAAAAAAAGAAAAG NAAANNAN\n",
      "CAAAAAAAAAAAAGAA\n",
      "PASS chr17 43104065 AAAAAGAAAAAAAAAAGAAAAG NAAAGAA\n",
      "AAAAAAAAAAAGAA\n",
      "PASS chr17 43104065 AAAAAGAAAAAAAAAAGAAAAG NAAAGNA\n",
      "AAAAAAAAAAAGAA\n",
      "PASS chr17 43104065 AAAAAGAAAAAAAAAAGAAAAG NAAANAA\n",
      "AAAAAAAAAAAGAA\n",
      "PASS chr17 43104065 AAAAAGAAAAAAAAAAGAAAAG NAANNNA\n",
      "AAAAAAAAAAAGAA\n",
      "PASS chr17 43104065 AAAAAGAAAAAAAAAAGAAAAG NANAGNA\n",
      "AAAAAAAAAAAGAA\n",
      "PASS chr17 43104065 AAAAAGAAAAAAAAAAGAAAAG NANANAA\n",
      "AAAAAAAAAAAGAA\n",
      "PASS chr17 43104065 AAAAAGAAAAAAAAAAGAAAAG NNAAGAA\n",
      "AAAAAAAAAAAGAA\n",
      "PASS chr17 43104066 AAAAGAAAAAAAAAAGAAAAG NAAGAA\n",
      "AAAAAAAAAGAA\n",
      "PASS chr17 43104066 AAAAGAAAAAAAAAAGAAAAG NAAGAN\n",
      "AAAAAAAAAGAA\n",
      "PASS chr17 43104066 AAAAGAAAAAAAAAAGAAAAG NAANAA\n",
      "AAAAAAAAAGAA\n",
      "PASS chr17 43104067 AAAGAAAAAAAAAAGAAAAG NAGAA\n",
      "AAAAAAAGAA\n",
      "PASS chr17 43104067 AAAGAAAAAAAAAAGAAAAG NAGAN\n",
      "AAAAAAAGAA\n",
      "PASS chr17 43104067 AAAGAAAAAAAAAAGAAAAG NAGNA\n",
      "AAAAAAAGAA\n",
      "PASS chr17 43104067 AAAGAAAAAAAAAAGAAAAG NANAA\n",
      "AAAAAAAGAA\n",
      "PASS chr17 43104067 AAAGAAAAAAAAAAGAAAAG NNGAA\n",
      "AAAAAAAGAA\n",
      "PASS chr17 43104068 AAGAAAAAAAAAAGAAAAG GAA\n",
      "AAAAGA\n",
      "PASS chr17 43104068 AAGAAAAAAAAAAGAAAAG NGAA\n",
      "AAAAAGAA\n",
      "PASS chr17 43104068 AAGAAAAAAAAAAGAAAAG NGNA\n",
      "AAAAAGAA\n",
      "PASS chr17 43104068 AAGAAAAAAAAAAGAAAAG NNAA\n",
      "AAAAAGAA\n",
      "FAIL  chr17 43104069 AGAAAAAAAAAAGAAAAG GAA\n",
      "AAAGAA\n",
      "PASS chr17 43104069 AGAAAAAAAAAAGAAAAG GNA\n",
      "AAAGAA\n",
      "PASS chr17 43104069 AGAAAAAAAAAAGAAAAG NAA\n",
      "AAAGAA\n",
      "PASS chr17 43104069 AGAAAAAAAAAAGAAAAG NNA\n",
      "AAAGAA\n",
      "FAIL  chr17 43104072 AAAAAAAAAGAAAAG A\n",
      "AA\n",
      "FAIL  chr17 43104083 AAAG A\n",
      "AA\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig 'chr16' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr17' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19' is not defined in the header. (Quick workaround: index the file with tabix.)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FAIL  chr17 58611365 G GA\n",
      "CGAA\n",
      "FAIL  chr19 1220729 G GC\n",
      "CGCC\n",
      "FAIL  chr19 4097205 TA T\n",
      "TA\n",
      "FAIL  chr19 4110475 GCCCTGC G\n",
      "GC\n",
      "FAIL  chr19 4117638 AC A\n",
      "AC\n",
      "FAIL  chr19 14093595 AACCCTCAGCAGGCTTAAGGAGGGGAGGCCCACCTTCCTCTGGTAG A\n",
      "AA\n",
      "FAIL  chr19 43543340 TCGTG T\n",
      "TC\n",
      "FAIL  chr19 43543341 C CGT\n",
      "ATCGTG\n",
      "FAIL  chr19 43543341 CGT C\n",
      "CG\n",
      "FAIL  chr19 43543341 CGTGT C\n",
      "CG\n",
      "FAIL  chr19 50401797 TG T\n",
      "TG\n",
      "FAIL  chr19 50414747 TG T\n",
      "TG\n",
      "FAIL  chr19 50416412 TG T\n",
      "TG\n",
      "FAIL  chr19 57235121 GCA G\n",
      "GC\n",
      "FAIL  chr20 59864399 G GA\n",
      "AGAA\n",
      "FAIL  chr20 59864457 GA G\n",
      "GA\n",
      "FAIL  chr20 59864457 G GA\n",
      "AGAA\n",
      "FAIL  chr22 41784278 GT G\n",
      "GT\n",
      "FAIL  chr22 41793806 A AT\n",
      "GATT\n",
      "FAIL  chr22 41793806 AT A\n",
      "AT\n",
      "FAIL  chr22 41795898 T TG\n",
      "TTGG\n",
      "FAIL  chrX 70678891 TG T\n",
      "TG\n",
      "FAIL  chrX 70678892 GAAA G\n",
      "GA\n",
      "FAIL  chrX 70678892 GAA G\n",
      "GA\n",
      "FAIL  chrX 70678892 GA G\n",
      "GA\n",
      "FAIL  chrX 70678934 GTCTCCCTCTCCCTCTCCC G\n",
      "GT\n",
      "PASS chrX 70678952 C CCACGG\n",
      "TCTCCCTCTCCC\n",
      "FAIL  chrX 70740803 GA G\n",
      "GA\n",
      "FAIL  chrX 70740803 G GA\n",
      "AGAA\n",
      "FAIL  chrX 70744234 GA G\n",
      "GA\n",
      "FAIL  chrX 70744241 A AT\n",
      "AAAT\n",
      "FAIL  chrX 133026979 A ATCTTCCCCAAATTCACTTTTAGGCTTCCAAATTGTGTTGAGTTTTTCCATGTTATCTTTCAGTTGATCTAAACAGTGAGCTAAAAACTCATGAGCATCGTTCTGTGCATTGCCATGGAATATCTCTGCAGCTGCTGAAATGGCCTTTTTAAGATTCAAGAGTAACATCTCCTTGATTTCTATATTATAGGTATCTTTAAAAAAAAGTAGCCGTGCCAAGCACATGGTAAGAGCATTAAGGGGAATTTTACCCCATGGGAAACTCTGATTAAGTAAATCATCAGCAAACGATGGGATTGAAAGTAGAGACTGTAACACTGCATTCATATAACAGGTGTTTCCCAAATTGGGGAGGCCGTGGCATATTTTCTCTGGAAATAATTCAAAAAATAGTTTTAATTTATCCCACTTTGTGTAACCGTCACTATACCCTTGTTGTAACAGAAATACCAAAACCATTTTCTCAGTGAGAGCTTGGAGAAGACCAATGTCATCTAGGTAAGGATTTCCAGTGGCGTTCATGATGCATGAAGATTCACATTCCAATTTCTTATTCTCTTCTAACTCTTTTAACTTCAATTGTTTCTCTCGATTATAGCTTACACACCTCGAACAATCTGCCTTGGATTTCTTGTATTCTACAGAATTATTTTCTTTCAAGAATTCCTCATTCATCTCTGAGCTAGATGAGAGCATTCTTTTCCTCTTCTTGTGCTGATTTTCTGATAACTCTCCGCAAGTAAG\n",
      "TCCAGATACGTCATTAAGTGAGCTAACGGTTCTTTTTCAATGAATGCTCGATCTCCTGAGTATACAGATTCTAGCTCATTTGGTTTAGAATTTTTTCCAAGGTACTTTTGCTGCTGTCTTCTGCTTGCCCCTTTAAAGACTGTCTGGCCTTTTTTTTGTTCAGACTCCTTATCTGATCCAATGTGTGGAGCCAGGATATCTTTGGATTCCTTTGTTGCAGGCCATGATACACTGATGTTTCCAGAAGTCATCTTTCGAATAACTTTTAATAATTGGAAATCTGTAATTTCTCCATCCTCACTCAAGGGAAGAGGTGGTCTGGTGCCTTCATTGCAATGAGAAGACACCTTTAAATATTTGGAAATGATGACTTCCTGGTCATTCTTCTTTAATGCACAAAACTCATTCAAGCTATAGCGTTTGAGGTGAACAATAAGGATTCTAGGTAGCCTACTGAATGAGTGCACTCCAACGGAAGTCTTGTGCTCACATTTTGCACATTTATACTCAAGCTCTTCTGCTCCAAAAAAAAGATCAAAAGTAGACTGAATAGATGAAGGATGTGCTTTTATTCTTTGGGGAAGGTTGATGGAGAGGTAATTATTCAGTTCTGTCTTGAGAATAACCTGACCACAAGCTTTACAAGCAATGGAGTGCAACAACTCTAACTCAAAATTAGTAATGACAGGGCAAGAAAACCCACTGGTGTCAGGATCATCAGCAAAAACCTGTTTAGGAAAATTATCTTCCCCAAATTCACTTTTAGGCTTCCAAATTGTGTTGAGTTTTTCCATGTTATCTTTCAGTTGATCTAAACAGTGAGCTAAAAACTCATGAGCATCGTTCTGTGCATTGCCATGGAATATCTCTGCAGCTGCTGAAATGGCCTTTTTAAGATTCAAGAGTAACATCTCCTTGATTTCTATATTATAGGTATCTTTAAAAAAAAGTAGCCGTGCCAAGCACATGGTAAGAGCATTAAGGGGAATTTTACCCCATGGGAAACTCTGATTAAGTAAATCATCAGCAAACGATGGGATTGAAAGTAGAGACTGTAACACTGCATTCATATAACAGGTGTTTCCCAAATTGGGGAGGCCGTGGCATATTTTCTCTGGAAATAATTCAAAAAATAGTTTTAATTTATCCCACTTTGTGTAACCGTCACTATACCCTTGTTGTAACAGAAATACCAAAACCATTTTCTCAGTGAGAGCTTGGAGAAGACCAATGTCATCTAGGTAAGGATTTCCAGTGGCGTTCATGATGCATGAAGATTCACATTCCAATTTCTTATTCTCTTCTAACTCTTTTAACTTCAATTGTTTCTCTCGATTATAGCTTACACACCTCGAACAATCTGCCTTGGATTTCTTGTATTCTACAGAATTATTTTCTTTCAAGAATTCCTCATTCATCTCTGAGCTAGATGAGAGCATTCTTTTCCTCTTCTTGTGCTGATTTTCTGATAACTCTCCGCAAGTAAGT\n",
      "FAIL  chrY 12833900 G GT\n",
      "TGTT\n",
      "FAIL  chrY 12837845 TA T\n",
      "TA\n",
      "FAIL  chrY 12837845 T TA\n",
      "TTAA\n",
      "FAIL  chrY 12840645 AG A\n",
      "AG\n",
      "FAIL  chrY 12842477 GC G\n",
      "GC\n",
      "FAIL  chrY 12842478 C CT\n",
      "GCTT\n",
      "FAIL  chrY 12842478 CT C\n",
      "CT\n",
      "FAIL  chrY 12842478 CTT C\n",
      "CT\n",
      "FAIL  chrY 12846292 G GT\n",
      "AGTT\n",
      "FAIL  chrY 12856902 AT A\n",
      "AT\n",
      "FAIL  chrY 12916485 CT C\n",
      "CT\n",
      "1SCS014: Passcount is 55 and failcount is 255\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig 'chr20' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr22' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrX' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrY' is not defined in the header. (Quick workaround: index the file with tabix.)\n"
     ]
    }
   ],
   "source": [
    "### 1 \n",
    "### GOOD TO GO  \n",
    "### Simplest code: is alternate allele in reference genome sequence twice the length of alternate allele\n",
    "\n",
    "# vcf = cyvcf2.VCF(sys.argv[2])\n",
    "vcf = VCF(A)\n",
    "\n",
    "failcount = 0\n",
    "passcount = 0 \n",
    "\n",
    "for v in vcf():\n",
    "    #lenalt = max(len(v.ALT[0]), len(v.REF))\n",
    "    lenalt = len(v.ALT[0])\n",
    "    genome_seq_stretch = genome[v.CHROM][v.POS-lenalt:v.POS+lenalt]\n",
    "    if v.ALT[0] in genome_seq_stretch:\n",
    "        print('FAIL ', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        print(genome_seq_stretch)\n",
    "        failcount += 1 \n",
    "    else:\n",
    "        print('PASS', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        print(genome_seq_stretch)\n",
    "        passcount += 1\n",
    "print(v.INFO.get('donor') + \":\", 'Passcount is', passcount, 'and failcount is', failcount) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### 2\n",
    "###  GOOD TO GO (I THINK)\n",
    "### More complex: let's think about allele size between reference and alternate alleles \n",
    "## Insertions: alt allele > ref allele\n",
    "## Version where there's repeat? Or I remove ref allele? \n",
    "## Look for alt allele in reference genome \n",
    "## Deletions: ref allele > alt allele \n",
    "## Is alt allele contained within reference allele? Is the string difference in the left or right stretch?\n",
    "\n",
    "# vcf = cyvcf2.VCF(sys.argv[2])\n",
    "vcf = VCF(A)\n",
    "\n",
    "failcount = 0\n",
    "passcount = 0 \n",
    "\n",
    "for v in vcf():\n",
    "    lenalt = max(len(v.ALT[0]), len(v.REF))\n",
    "    genome_seq_left = genome[v.CHROM][v.POS-lenalt:v.POS]\n",
    "    genome_seq_right = genome[v.CHROM][v.POS:v.POS+lenalt]\n",
    "    genome_seq_stretch = genome[v.CHROM][v.POS-lenalt:v.POS+lenalt]\n",
    "    if len(v.ALT[0]) > len(v.REF) and v.ALT[0] in genome_seq_stretch:\n",
    "        print('FAIL ', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "        failcount += 1 \n",
    "    elif len(v.ALT[0]) < len(v.REF):\n",
    "        valt = v.ALT[0]\n",
    "        vref = v.REF\n",
    "        if valt in vref:\n",
    "            vref = vref.replace(valt, '', 1)\n",
    "            #only replace once \n",
    "            if vref in genome_seq_left or genome_seq_right:\n",
    "                print('FAIL ', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "                print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "                failcount += 1 \n",
    "            else: \n",
    "                print('PASS ', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "                print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "                passcount += 1\n",
    "        else:\n",
    "            print('PASS ', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "            print(genome_seq_left, genome_seq_right, genome_seq_stretch)    \n",
    "            passcount += 1\n",
    "    else:\n",
    "        print('PASS', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "        passcount += 1\n",
    "print(v.INFO.get('donor') + \":\", 'Passcount is', passcount, 'and failcount is', failcount) \n",
    "   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### 3\n",
    "###  Probably good to go \n",
    "### Less strict: checks for larger alt alleles in genome stretch and only homopolymers fail with regards to reference allele in adjacent stretches\n",
    "vcf = VCF(A)\n",
    "\n",
    "failcount = 0\n",
    "passcount = 0 \n",
    "\n",
    "for v in vcf():\n",
    "    lenalt = max(len(v.ALT[0]), len(v.REF))\n",
    "    genome_seq_left = genome[v.CHROM][v.POS-lenalt:v.POS]\n",
    "\n",
    "    if (genome_seq_left == len(genome_seq_left) * genome_seq_left[0]):\n",
    "        genome_seq_left_use = genome_seq_left\n",
    "    else:\n",
    "        genome_seq_left_use = 'no_left'\n",
    "    genome_seq_right = genome[v.CHROM][v.POS:v.POS+lenalt]\n",
    "    if (genome_seq_right == len(genome_seq_right) * genome_seq_right[0]):\n",
    "        genome_seq_right_use = genome_seq_right\n",
    "    else:\n",
    "        genome_seq_right_use = 'no_right'\n",
    "    genome_seq_stretch = genome[v.CHROM][v.POS-lenalt:v.POS+lenalt]\n",
    "\n",
    "    if len(v.ALT[0]) > len(v.REF) and v.ALT[0] in genome_seq_stretch:\n",
    "        print('FAIL in stretch', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "        failcount += 1 \n",
    "    elif (len(v.ALT[0]) > len(v.REF)) and (v.ALT[0] not in genome_seq_stretch):\n",
    "        print('PASS should not be in stretch', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "        passcount += 1\n",
    "    elif len(v.ALT[0]) < len(v.REF):\n",
    "        valt = v.ALT[0]\n",
    "        vref = v.REF\n",
    "        if vref in valt:\n",
    "            vref = vref.replace(valt, '', 1)\n",
    "            if vref in genome_seq_left_use or genome_seq_right_use:\n",
    "                print('FAIL right or left', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "                print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "                failcount += 1 \n",
    "            else: \n",
    "                print('PASS not in left or right', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "                print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "                passcount += 1\n",
    "\n",
    "        else:\n",
    "            print('PASS leftover', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "            print(genome_seq_left, genome_seq_right, genome_seq_stretch)    \n",
    "            passcount += 1\n",
    "    else:\n",
    "        print('PASS leftover', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "        passcount += 1\n",
    "print(v.INFO.get('donor') + \":\", 'Passcount is', passcount, 'and failcount is', failcount) \n",
    "   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### 4, fine but could be fiddled with \n",
    "### Least strict: only homopolymers fail with regards to reference allele in adjacent stretches AND alternate alleles in homopolymers\n",
    "vcf = VCF('1SCS014.trustworthyindels.vcf')\n",
    "\n",
    "failcount = 0\n",
    "passcount = 0 \n",
    "\n",
    "for v in vcf():\n",
    "    lenalt = max(len(v.ALT[0]), len(v.REF))\n",
    "    genome_seq_left = genome[v.CHROM][v.POS-lenalt:v.POS]\n",
    "    if (genome_seq_left == len(genome_seq_left) * genome_seq_left[0]):\n",
    "        genome_seq_left_use = genome_seq_left\n",
    "    else:\n",
    "        genome_seq_left_use = 'Left'\n",
    "    genome_seq_right = genome[v.CHROM][v.POS:v.POS+lenalt]\n",
    "    if (genome_seq_right == len(genome_seq_right) * genome_seq_right[0]):\n",
    "        genome_seq_right_use = genome_seq_right\n",
    "    else:\n",
    "        genome_seq_right_use = 'Right'\n",
    "    genome_seq_stretch = genome[v.CHROM][v.POS-lenalt:v.POS+lenalt]\n",
    "    if (genome_seq_stretch == len(genome_seq_stretch) * genome_seq_stretch[0]):\n",
    "        genome_seq_stretch_use = genome_seq_stretch\n",
    "    else:\n",
    "        genome_seq_stretch_use = 'Stretch'\n",
    "    valt = v.ALT[0]\n",
    "    vref = v.REF\n",
    "    if len(v.ALT[0]) < len(v.REF):\n",
    "        vref = vref.replace(valt, '', 1)\n",
    "    elif len(v.ALT[0]) > len(v.REF):\n",
    "        valt = valt.replace(vref, '', 1)\n",
    "    else:\n",
    "        pass\n",
    "\n",
    "    if len(v.ALT[0]) > len(v.REF) and valt in genome_seq_stretch_use:\n",
    "        #print('FAIL in stretch', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        #print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "        failcount += 1 \n",
    "    elif (len(v.ALT[0]) > len(v.REF)) and (valt not in genome_seq_stretch_use):\n",
    "        #print('PASS not in stretch/stretch not homopolymer', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        #print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "        passcount += 1\n",
    "    elif len(v.ALT[0]) < len(v.REF):\n",
    "        if vref in valt:\n",
    "            vref = vref.replace(valt, '', 1)\n",
    "            if vref in genome_seq_left_use or genome_seq_right_use:\n",
    "                #print('FAIL right or left', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "                #print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "                failcount += 1 \n",
    "            else: \n",
    "                #print('PASS not in left or right', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "                #print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "                passcount += 1\n",
    "        else:\n",
    "            #print('PASS leftover', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "            #print(genome_seq_left, genome_seq_right, genome_seq_stretch)    \n",
    "            passcount += 1\n",
    "    else:\n",
    "        #print('PASS leftover', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        #print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "        passcount += 1\n",
    "print(v.INFO.get('donor') + \":\", 'Passcount is', passcount, 'and failcount is', failcount) \n",
    "   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Substring(s):\n",
    " \n",
    "    ans, temp = 1, 1\n",
    " \n",
    "    # Traverse the string\n",
    "    for i in range(1, len(s)):\n",
    "         \n",
    "        # If character is same as\n",
    "        # previous increment temp value\n",
    "        if (s[i] == s[i - 1]):\n",
    "            temp += 1\n",
    "        else:\n",
    "            ans = max(ans, temp)\n",
    "            temp = 1\n",
    " \n",
    "    ans = max(ans, temp)\n",
    " \n",
    "    # Return the required answer\n",
    "    return ans\n",
    " \n",
    "# Driver code\n",
    "s = \"abcdddddeff\"\n",
    " \n",
    "# Function call\n",
    "print(Substring(s))\n",
    "# Mohit Kumar "
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
  },
  "kernelspec": {
   "display_name": "Python 3.8.9 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
