{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Running questions\n",
    "### How to handle Ns? [We're probably going to ignore them]\n",
    "### Should I remove reference from alternate the same way I remove alternate from the reference? \n",
    "### Find longest string rather than checking entire stretch and neighbors to see if they're homopolymers? \n",
    "\n",
    "### Note: I will clean up stuff for sure (moving lines to functions, etc), this is more about logic "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cyvcf2\n",
    "from pyfaidx import Fasta\n",
    "import os\n",
    "import argparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/Users/quinlan/u1264408/Git/spermseq')\n",
    "os.listdir()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A = '1SCS014.indels.vcf'\n",
    "Atrust = '1SCS014.trustworthyindels.vcf'\n",
    "B = 'K8J7003.indels.vcf'\n",
    "C = 'UA6U012.indels.vcf'\n",
    "D = 'YL99029.indels.vcf'\n",
    "E = 'XH9R013.indels.vcf'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = Fasta('/Users/quinlan/u1264408/human_g1k_v38_decoy_phix.fasta', as_raw=True) # look into as_RAW\n",
    "# as_raw: optional parameter to specify whether to return sequences as a Sequence() object or as a raw string.\n",
    "from cyvcf2 import VCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### 1 \n",
    "### GOOD TO GO  \n",
    "### Simplest code: is alternate allele in reference genome sequence twice the length of alternate allele\n",
    "\n",
    "# vcf = cyvcf2.VCF(sys.argv[2])\n",
    "vcf = VCF(Atrust)\n",
    "\n",
    "failcount = 0\n",
    "passcount = 0 \n",
    "\n",
    "for v in vcf():\n",
    "    #lenalt = max(len(v.ALT[0]), len(v.REF))\n",
    "    lenalt = len(v.ALT[0])\n",
    "    genome_seq_stretch = genome[v.CHROM][v.POS-lenalt:v.POS+lenalt]\n",
    "    if 'N' in v.ALT[0]:\n",
    "        #print('FAIL ', v.ALT[0])\n",
    "        failcount += 1 \n",
    "    elif v.ALT[0] in genome_seq_stretch:\n",
    "        #print('FAIL ', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        #print(genome_seq_stretch)\n",
    "        failcount += 1 \n",
    "    else:\n",
    "        #print('PASS', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        #print(genome_seq_stretch)\n",
    "        passcount += 1\n",
    "print(v.INFO.get('donor') + \":\", 'Passcount is', passcount, 'and failcount is', failcount, \".\", \"Total is\", (passcount + failcount)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### 2\n",
    "###  GOOD TO GO (I THINK)\n",
    "### More complex: let's think about allele size between reference and alternate alleles \n",
    "## Insertions: alt allele > ref allele\n",
    "## Version where there's repeat? Or I remove ref allele? \n",
    "## Look for alt allele in reference genome \n",
    "## Deletions: ref allele > alt allele \n",
    "## Is alt allele contained within reference allele? Is the string difference in the left or right stretch?\n",
    "\n",
    "# vcf = cyvcf2.VCF(sys.argv[2])\n",
    "vcf = VCF(Atrust)\n",
    "\n",
    "failcount = 0\n",
    "passcount = 0 \n",
    "\n",
    "for v in vcf():\n",
    "    lenalt = max(len(v.ALT[0]), len(v.REF))\n",
    "    genome_seq_left = genome[v.CHROM][v.POS-lenalt:v.POS]\n",
    "    genome_seq_right = genome[v.CHROM][v.POS:v.POS+lenalt]\n",
    "    genome_seq_stretch = genome[v.CHROM][v.POS-lenalt:v.POS+lenalt]\n",
    "    if 'N' in v.ALT[0]:\n",
    "        print('FAIL ', v.ALT[0])\n",
    "        failcount += 1 \n",
    "    elif len(v.ALT[0]) > len(v.REF) and v.ALT[0] in genome_seq_stretch:\n",
    "        print('FAIL ', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "        failcount += 1 \n",
    "    elif len(v.ALT[0]) < len(v.REF):\n",
    "        valt = v.ALT[0]\n",
    "        vref = v.REF\n",
    "        if valt in vref:\n",
    "            vref = vref.replace(valt, '', 1)\n",
    "            #only replace once \n",
    "            if vref in genome_seq_left or genome_seq_right:\n",
    "                print('FAIL ', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "                print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "                failcount += 1 \n",
    "            else: \n",
    "                print('PASS ', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "                print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "                passcount += 1\n",
    "        else:\n",
    "            print('PASS ', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "            print(genome_seq_left, genome_seq_right, genome_seq_stretch)    \n",
    "            passcount += 1\n",
    "    else:\n",
    "        print('PASS', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "        passcount += 1\n",
    "print(v.INFO.get('donor') + \":\", 'Passcount is', passcount, 'and failcount is', failcount) \n",
    "   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### 3\n",
    "###  Probably good to go \n",
    "### Less strict: checks for larger alt alleles in genome stretch and only homopolymers fail with regards to reference allele in adjacent stretches\n",
    "vcf = VCF(Atrust)\n",
    "\n",
    "failcount = 0\n",
    "passcount = 0 \n",
    "\n",
    "for v in vcf():\n",
    "    lenalt = max(len(v.ALT[0]), len(v.REF))\n",
    "    genome_seq_left = genome[v.CHROM][v.POS-lenalt:v.POS]\n",
    "    genome_seq_right = genome[v.CHROM][v.POS:v.POS+lenalt]\n",
    "    \n",
    "    if (genome_seq_left == len(genome_seq_left) * genome_seq_left[0]):\n",
    "        genome_seq_left_use = genome_seq_left\n",
    "    else:\n",
    "        genome_seq_left_use = 'no_left'\n",
    "    genome_seq_right = genome[v.CHROM][v.POS:v.POS+lenalt]\n",
    "    if (genome_seq_right == len(genome_seq_right) * genome_seq_right[0]):\n",
    "        genome_seq_right_use = genome_seq_right\n",
    "    else:\n",
    "        genome_seq_right_use = 'no_right'\n",
    "    genome_seq_stretch = genome[v.CHROM][v.POS-lenalt:v.POS+lenalt]\n",
    "\n",
    "    if 'N' in v.ALT[0]:\n",
    "        print('FAIL ', v.ALT[0])\n",
    "        failcount += 1 \n",
    "    elif len(v.ALT[0]) > len(v.REF) and v.ALT[0] in genome_seq_stretch:\n",
    "        print('FAIL in stretch', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "        failcount += 1 \n",
    "    elif (len(v.ALT[0]) > len(v.REF)) and (v.ALT[0] not in genome_seq_stretch):\n",
    "        print('PASS should not be in stretch', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "        passcount += 1\n",
    "    elif len(v.ALT[0]) < len(v.REF):\n",
    "        valt = v.ALT[0]\n",
    "        vref = v.REF\n",
    "        if vref in valt:\n",
    "            vref = vref.replace(valt, '', 1)\n",
    "            if vref in genome_seq_left_use or genome_seq_right_use:\n",
    "                print('FAIL right or left', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "                print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "                failcount += 1 \n",
    "            else: \n",
    "                print('PASS not in left or right', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "                print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "                passcount += 1\n",
    "\n",
    "        else:\n",
    "            print('PASS leftover', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "            print(genome_seq_left, genome_seq_right, genome_seq_stretch)    \n",
    "            passcount += 1\n",
    "    else:\n",
    "        print('PASS leftover', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "        passcount += 1\n",
    "print(v.INFO.get('donor') + \":\", 'Passcount is', passcount, 'and failcount is', failcount) \n",
    "   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### 4, fine but could be fiddled with \n",
    "### Least strict: only homopolymers fail with regards to reference allele in adjacent stretches AND alternate alleles in homopolymers\n",
    "vcf = VCF(Atrust)\n",
    "\n",
    "failcount = 0\n",
    "passcount = 0 \n",
    "\n",
    "for v in vcf():\n",
    "    lenalt = max(len(v.ALT[0]), len(v.REF))\n",
    "    genome_seq_left = genome[v.CHROM][v.POS-lenalt:v.POS]\n",
    "    genome_seq_right = genome[v.CHROM][v.POS:v.POS+lenalt]\n",
    "    genome_seq_stretch = genome[v.CHROM][v.POS-lenalt:v.POS+lenalt]\n",
    "    \n",
    "    if (genome_seq_left == len(genome_seq_left) * genome_seq_left[0]):\n",
    "        genome_seq_left_use = genome_seq_left\n",
    "    else:\n",
    "        genome_seq_left_use = 'Left'\n",
    "    genome_seq_right = genome[v.CHROM][v.POS:v.POS+lenalt]\n",
    "    if (genome_seq_right == len(genome_seq_right) * genome_seq_right[0]):\n",
    "        genome_seq_right_use = genome_seq_right\n",
    "    else:\n",
    "        genome_seq_right_use = 'Right'\n",
    "    genome_seq_stretch = genome[v.CHROM][v.POS-lenalt:v.POS+lenalt]\n",
    "    if (genome_seq_stretch == len(genome_seq_stretch) * genome_seq_stretch[0]):\n",
    "        genome_seq_stretch_use = genome_seq_stretch\n",
    "    else:\n",
    "        genome_seq_stretch_use = 'Stretch'\n",
    "    valt = v.ALT[0]\n",
    "    vref = v.REF\n",
    "    if len(v.ALT[0]) < len(v.REF):\n",
    "        vref = vref.replace(valt, '', 1)\n",
    "    elif len(v.ALT[0]) > len(v.REF):\n",
    "        valt = valt.replace(vref, '', 1)\n",
    "    else:\n",
    "        pass\n",
    "\n",
    "    if 'N' in v.ALT[0]:\n",
    "        print('FAIL ', v.ALT[0])\n",
    "        failcount += 1 \n",
    "    elif len(v.ALT[0]) > len(v.REF) and valt in genome_seq_stretch_use:\n",
    "        #print('FAIL in stretch', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        #print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "        failcount += 1 \n",
    "    elif (len(v.ALT[0]) > len(v.REF)) and (valt not in genome_seq_stretch_use):\n",
    "        #print('PASS not in stretch/stretch not homopolymer', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        #print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "        passcount += 1\n",
    "    elif len(v.ALT[0]) < len(v.REF):\n",
    "        if vref in valt:\n",
    "            vref = vref.replace(valt, '', 1)\n",
    "            if vref in genome_seq_left_use or genome_seq_right_use:\n",
    "                #print('FAIL right or left', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "                #print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "                failcount += 1 \n",
    "            else: \n",
    "                #print('PASS not in left or right', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "                #print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "                passcount += 1\n",
    "        else:\n",
    "            #print('PASS leftover', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "            #print(genome_seq_left, genome_seq_right, genome_seq_stretch)    \n",
    "            passcount += 1\n",
    "    else:\n",
    "        #print('PASS leftover', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        #print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "        passcount += 1\n",
    "print(v.INFO.get('donor') + \":\", 'Passcount is', passcount, 'and failcount is', failcount) \n",
    "   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Substring(s):\n",
    " \n",
    "    ans, temp = 1, 1\n",
    " \n",
    "    # Traverse the string\n",
    "    for i in range(1, len(s)):\n",
    "         \n",
    "        # If character is same as\n",
    "        # previous increment temp value\n",
    "        if (s[i] == s[i - 1]):\n",
    "            temp += 1\n",
    "        else:\n",
    "            ans = max(ans, temp)\n",
    "            temp = 1\n",
    " \n",
    "    ans = max(ans, temp)\n",
    " \n",
    "    # Return the required answer\n",
    "    return ans\n",
    " \n",
    "# Driver code\n",
    "s = \"abcdddddeff\"\n",
    " \n",
    "# Function call\n",
    "print(Substring(s))\n",
    "# Mohit Kumar "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Strategy 5: a modified strategy 3\n",
    "# taking only the difference between v.alt and v.ref in analysis\n",
    "\n",
    "vcf = VCF(A)\n",
    "\n",
    "failcount = 0\n",
    "passcount = 0 \n",
    "\n",
    "for v in vcf():\n",
    "    lenalt = max(len(v.ALT[0]), len(v.REF))\n",
    "    genome_seq_left = genome[v.CHROM][v.POS-lenalt:v.POS]\n",
    "    genome_seq_right = genome[v.CHROM][v.POS:v.POS+lenalt]\n",
    "    \n",
    "    if (genome_seq_left == len(genome_seq_left) * genome_seq_left[0]):\n",
    "        genome_seq_left_use = genome_seq_left\n",
    "    else:\n",
    "        genome_seq_left_use = 'no_left'\n",
    "    genome_seq_right = genome[v.CHROM][v.POS:v.POS+lenalt]\n",
    "    if (genome_seq_right == len(genome_seq_right) * genome_seq_right[0]):\n",
    "        genome_seq_right_use = genome_seq_right\n",
    "    else:\n",
    "        genome_seq_right_use = 'no_right'\n",
    "    genome_seq_stretch = genome[v.CHROM][v.POS-lenalt:v.POS+lenalt]\n",
    "\n",
    "    valt = v.ALT[0]\n",
    "    vref = v.REF\n",
    "    if len(vref) > len(valt):\n",
    "        vref = vref.replace(valt, '', 1)\n",
    "    else: \n",
    "        valt = valt.replace(vref, '', 1)\n",
    "\n",
    "    if 'N' in v.ALT[0]:\n",
    "        print('FAIL ', v.ALT[0])\n",
    "        failcount += 1 \n",
    "    elif len(v.ALT[0]) > len(v.REF) and valt in genome_seq_stretch:\n",
    "        print('FAIL in stretch', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "        failcount += 1 \n",
    "    elif (len(v.ALT[0]) > len(v.REF)) and (valt not in genome_seq_stretch):\n",
    "        print('PASS should not be in stretch', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "        passcount += 1\n",
    "    elif len(v.ALT[0]) < len(v.REF):\n",
    "        if vref in genome_seq_left_use or genome_seq_right_use:\n",
    "            print('FAIL right or left', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "            print(genome_seq_left, genome_seq_right, genome_seq_stretch)                \n",
    "            failcount += 1 \n",
    "        else: \n",
    "            print('PASS not in left or right', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "            print(genome_seq_left, genome_seq_right, genome_seq_stretch)                \n",
    "            passcount += 1\n",
    "         \n",
    "    else:\n",
    "        print('PASS leftover', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "        passcount += 1\n",
    "print(v.INFO.get('donor') + \":\", 'Passcount is', passcount, 'and failcount is', failcount) \n",
    "   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### 3 BUT in left and right instead of stretch\n",
    "###  Probably good to go \n",
    "### Less strict: checks for larger alt alleles in genome stretch and only homopolymers fail with regards to reference allele in adjacent stretches\n",
    "vcf = VCF(Atrust)\n",
    "\n",
    "failcount = 0\n",
    "passcount = 0 \n",
    "\n",
    "for v in vcf():\n",
    "    lenalt = max(len(v.ALT[0]), len(v.REF))\n",
    "    genome_seq_left = genome[v.CHROM][v.POS-lenalt:v.POS]\n",
    "    genome_seq_right = genome[v.CHROM][v.POS:v.POS+lenalt]\n",
    "    \n",
    "    if (genome_seq_left == len(genome_seq_left) * genome_seq_left[0]):\n",
    "        genome_seq_left_use = genome_seq_left\n",
    "    else:\n",
    "        genome_seq_left_use = 'no_left'\n",
    "    genome_seq_right = genome[v.CHROM][v.POS:v.POS+lenalt]\n",
    "    if (genome_seq_right == len(genome_seq_right) * genome_seq_right[0]):\n",
    "        genome_seq_right_use = genome_seq_right\n",
    "    else:\n",
    "        genome_seq_right_use = 'no_right'\n",
    "    genome_seq_stretch = genome[v.CHROM][v.POS-lenalt:v.POS+lenalt]\n",
    "    \n",
    "    valt = v.ALT[0]\n",
    "    vref = v.REF\n",
    "    if len(vref) > len(valt):\n",
    "        vref = vref.replace(valt, '', 1)\n",
    "    else: \n",
    "        valt = valt.replace(vref, '', 1)\n",
    "    if 'N' in v.ALT[0]:\n",
    "        print('FAIL ', v.ALT[0])\n",
    "        failcount += 1 \n",
    "    elif len(v.ALT[0]) > len(v.REF) and (valt in (genome_seq_left or genome_seq_right)):\n",
    "        print('FAIL in stretch', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "        failcount += 1 \n",
    "    elif (len(v.ALT[0]) > len(v.REF)) and (valt not in (genome_seq_left or genome_seq_right)):\n",
    "        print('PASS should not be in stretch', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "        passcount += 1\n",
    "    elif len(v.ALT[0]) < len(v.REF):\n",
    "        valt = v.ALT[0]\n",
    "        vref = v.REF\n",
    "        if vref in valt:\n",
    "            vref = vref.replace(valt, '', 1)\n",
    "            if vref in genome_seq_left_use or genome_seq_right_use:\n",
    "                print('FAIL right or left', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "                print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "                failcount += 1 \n",
    "            else: \n",
    "                print('PASS not in left or right', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "                print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "                passcount += 1\n",
    "\n",
    "        else:\n",
    "            print('PASS leftover', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "            print(genome_seq_left, genome_seq_right, genome_seq_stretch)    \n",
    "            passcount += 1\n",
    "    else:\n",
    "        print('PASS leftover', v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        print(genome_seq_left, genome_seq_right, genome_seq_stretch)\n",
    "        passcount += 1\n",
    "print(v.INFO.get('donor') + \":\", 'Passcount is', passcount, 'and failcount is', failcount) \n",
    "   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Substring(s):\n",
    "    ans, temp = 1, 1\n",
    "    # Traverse the string\n",
    "    for i in range(1, len(s)):\n",
    "        # If character is same as\n",
    "        # previous increment temp value\n",
    "        if (s[i] == s[i - 1]):\n",
    "            temp += 1\n",
    "        else:\n",
    "            ans = max(ans, temp)\n",
    "            temp = 1\n",
    " \n",
    "        ans = max(ans, temp)\n",
    "    # the required answer\n",
    "    return ans\n",
    "\n",
    "vcf = VCF(Atrust)\n",
    "\n",
    "failcount = 0\n",
    "passcount = 0 \n",
    "\n",
    "for v in vcf():\n",
    "    genome_seq_left = genome[v.CHROM][v.POS-10:v.POS]\n",
    "    genome_seq_right = genome[v.CHROM][v.POS:v.POS+10]\n",
    "    \n",
    "    valt = v.ALT[0]\n",
    "    vref = v.REF\n",
    "    if len(vref) > len(valt):\n",
    "        vdiff = vref.replace(valt, '', 1)\n",
    "    else: \n",
    "        vdiff = valt.replace(vref, '', 1)\n",
    "    \n",
    "    sleft = (Substring(genome_seq_left)\n",
    "    vlen = (len(vdiff))\n",
    "\n",
    "    print(Substring(genome_seq_left), genome_seq_left, Substring(genome_seq_right), genome_seq_right)\n",
    "    if (vdiff * )) in genome_seq_left:\n",
    "        print( v.CHROM, v.POS, v.REF, v.ALT[0])            \n",
    "        failcount += 1\n",
    "        print('vdiff in left')\n",
    "    elif (vdiff * (Substring(genome_seq_right)/len(vdiff))) in genome_seq_right: \n",
    "        print( v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        failcount += 1\n",
    "        print('vdiff in right')\n",
    "    else:\n",
    "        print( v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        passcount += 1\n",
    "        print('PASS')\n",
    "print(v.INFO.get('donor') + \":\", 'Passcount is', passcount, 'and failcount is', failcount) \n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8 CTAGTTTTTTTT 7 AAAAAAAGAAAT\n",
      "chr1 75806922 T TA\n",
      "vdiff in right\n",
      "3 ACATCTGGGAAT 9 AAAAAAAAAGTG\n",
      "chr2 39007201 T TA\n",
      "vdiff in right\n",
      "3 AAAGTGATTTAA 10 TTTTTTTTTTAA\n",
      "chr2 39035331 AT A\n",
      "vdiff in left\n",
      "5 CCTTTTTCAGGT 12 AAAAAAAAAAAA\n",
      "chr2 47414420 TAAA T\n",
      "PASS\n",
      "5 CCTTTTTCAGGT 12 AAAAAAAAAAAA\n",
      "chr2 47414420 TAA T\n",
      "PASS\n",
      "3 AAAGTGTAAATT 3 AGGAAATGGGTT\n",
      "chr2 47474913 T TAGG\n",
      "PASS\n",
      "4 TCATAAAAGACC 4 TTTTCCTCCCTC\n",
      "chr2 47804889 CT C\n",
      "vdiff in right\n",
      "3 AGTATGCTTTAG 10 AAAAAAAAAATT\n",
      "chr2 189852791 GA G\n",
      "vdiff in right\n",
      "2 GAAGCACAAGTT 5 AAAAACCTCATT\n",
      "chr2 189863889 TA T\n",
      "vdiff in left\n",
      "3 TAGAAAGGATTT 8 AAAAAAAACATG\n",
      "chr3 12609348 TA T\n",
      "vdiff in left\n",
      "2 GGATGAGGAGGT 3 AGAGCATGACCC\n",
      "chr4 54731256 T TAG\n",
      "PASS\n",
      "3 AATGTTAGGGAC 8 AAAAAAAAGGGC\n",
      "chr5 80675095 C CA\n",
      "vdiff in right\n",
      "3 AATAGCAGAAAT 4 AAAAGAAAAGAT\n",
      "chr5 112839514 TAAAAG T\n",
      "PASS\n",
      "2 AACTTCACCAGT 4 AAAACCTATACC\n",
      "chr5 112840921 TA T\n",
      "vdiff in left\n",
      "6 ATGCCAAAAAAG 5 AAAAAGCCTTCA\n",
      "chr5 112841744 GA G\n",
      "vdiff in left\n",
      "5 AAAAATGTCATA 2 TACATCTCCAGG\n",
      "chr5 112842691 A ATACATCTCCAGGTAGACAGATGAGCCAACAGAACCTTACCAAACAAACAGGTTTATCCAAGAATGCCAGTAGTATTCCAAGAAGTGAGTCTGCCTCCAAAGGACTAAATCAGATGAATAATGGTAATGGAGCCAATAAAAAGGTAGAACTTTCTAGAATGTCTTCAACTAAATCAAGTGGAAGTGAATCTGATAGATCAGAAAGACCTGTATTAGTACGCCAGTCAACTTTCATCAAAGAAGCTCCAAGCCCAACCTTAAGAAGAAAATTGGAGGAATCTGCTTCATTTGAATCTCTTTCTCCATCATCTAGACCAGCTTCTCCCACTAGGTCCCAGGCACAAACTCCAGTTTTAAG\n",
      "PASS\n",
      "2 CTCCAACGGCCT 6 GGGGGGTTTCCC\n",
      "chr6 151808193 TG T\n",
      "vdiff in left\n",
      "7 TACTGTTTTTTT 6 CCCCCCAGGCCA\n",
      "chr6 151842588 TC T\n",
      "vdiff in right\n",
      "3 AAATGTTCCCAG 6 CCCCCCGCATTC\n",
      "chr7 5992106 GC G\n",
      "vdiff in left\n",
      "3 ATCACTATCTTT 9 AAAAAAAAAGCT\n",
      "chr7 5997264 TA T\n",
      "vdiff in right\n",
      "2 GCGAGAATCTAC 11 AAAAAAAAAAAG\n",
      "chr7 140734774 CA C\n",
      "vdiff in left\n",
      "2 GCGAGAATCTAC 11 AAAAAAAAAAAG\n",
      "chr7 140734774 C CA\n",
      "vdiff in left\n",
      "2 GGTAATGAGGCA 7 GGGGGGGTAGCA\n",
      "chr7 140783126 AG A\n",
      "vdiff in left\n",
      "3 AAAGCTTCAGTT 2 ATAATGCTAGCA\n",
      "chr8 30847884 TATA T\n",
      "PASS\n",
      "3 CAAAGAGAACCT 12 AAAAAAAAAAAA\n",
      "chr8 30860059 T TA\n",
      "vdiff in left\n",
      "4 GCTGCGGCCCCG 4 CCCAGGCCCCGC\n",
      "chr9 124482702 G GC\n",
      "vdiff in left\n",
      "3 GATGAAAGAGAT 9 AAAAAAAAACTC\n",
      "chr11 94486159 T TA\n",
      "vdiff in left\n",
      "2 TACCTATATGTA 8 TTTTTTTTACAG\n",
      "chr11 108227582 A AT\n",
      "vdiff in right\n",
      "2 TACCTATATGTA 8 TTTTTTTTACAG\n",
      "chr11 108227582 AT A\n",
      "vdiff in right\n",
      "2 GAACTTACATAG 12 TTTTTTTTTTTT\n",
      "chr11 108325249 GT G\n",
      "vdiff in left\n",
      "3 TGCAGCGGTTTG 8 TTTTTTTTATTG\n",
      "chr11 108326230 G GT\n",
      "vdiff in left\n",
      "4 GATGTCCCCGCA 6 CCCCCCTCTATT\n",
      "chr11 116835717 AC A\n",
      "vdiff in left\n",
      "2 GCCTTCCAGCCG 2 CACCACCACCAC\n",
      "chr11 119206522 GCAC G\n",
      "vdiff in right\n",
      "3 TTTCCATCATAA 9 TTTTTTTTTGCA\n",
      "chr12 101729122 AT A\n",
      "vdiff in left\n",
      "3 AAATGTGGCTCT 9 AAAAAAAAAGCA\n",
      "chr12 101734904 T TA\n",
      "vdiff in left\n",
      "3 TTTCAGGAGTTA 2 TCGTTAGCAGAG\n",
      "chr12 130345784 AT A\n",
      "vdiff in left\n",
      "3 AGGTTATTTGGG 4 AAAAGGGAGATG\n",
      "chr12 130346594 GA G\n",
      "vdiff in right\n",
      "4 AGGCGGGGCCCT 6 GGGGGGACACTG\n",
      "chr12 130354608 TG T\n",
      "vdiff in left\n",
      "2 CCTCTCCGTGAT 6 GGGGGGAGCCCT\n",
      "chr12 132659279 TG T\n",
      "vdiff in right\n",
      "2 ATGTTGAAGATC 7 AAAAAAACACTA\n",
      "chr13 32339699 CA C\n",
      "vdiff in left\n",
      "3 CAAAGGCAACGC 3 GTCTTTCCACAG\n",
      "chr13 32356498 C CGTCTTTCCACAGCCAGGCAGTCTGTATCTTGCAAAAACATCCACTCTGCCTCGAATCTCTCTGAAAGCAGCAGTAGGAGGCCAAGTTCCCTCTGCGTGTTCTCATAAACAGGTATGTGTTTGTCTACAATACTGATGGCTTTTATGACAGAGTGTAATTTTATTTCATTAACTAGTATCTACAAATGGCTTTGTTTAAAGAATGAACACATTAGTGCAGGAATGGATGAATGAAATCATCATATTTTCTAATTAGCCTGCAGTGGCAGCCTCTGGCCCCTTGCTAGGCCTGCCTCATCCTGCTAAAGTGATCTGTGCTTCCAAATTACTACTTCTTTTCCCCCTTCAAATCTTTCTTATTTTGTCATTGTAAATGCTCTCAGCTAGGTGTTAAAGTAGTCTTACTGATATTCAAATGTGAATAACTGATAGCCCTGAACCTTCTATGAGCTATTTATATTTTCCAAAGAGGATTCTCCTTAAGCCAATATTATCTAGGTAGAATTTTAGGCAATGGAGAGGTGAAAATAATATTGATGACATTAATAGCTAACTTTGAGCATTTTCTAG\n",
      "PASS\n",
      "4 AATTAAAGTTTT 1 TATATATATATA\n",
      "chr14 45159080 TTA T\n",
      "vdiff in right\n",
      "2 ATTGCTAGATTC 9 TTTTTTTTTCTC\n",
      "chr14 75047900 C CT\n",
      "vdiff in left\n",
      "2 AGAACACCGTCC 5 TTTTTCCATTTC\n",
      "chr14 95132538 CT C\n",
      "vdiff in right\n",
      "2 AATGGACATAAC 9 TTTTTTTTTCTC\n",
      "chr14 103966662 CT C\n",
      "vdiff in right\n",
      "3 CTTACATGCCCA 4 GGGGACTATTGA\n",
      "chr15 89307731 A AG\n",
      "vdiff in right\n",
      "4 ACAGCAATCCCC 4 AAAATGGGTAGC\n",
      "chr16 89795865 CA C\n",
      "vdiff in right\n",
      "3 ATTTGGTTGTAC 8 TTTTTTTTCTTT\n",
      "chr17 43093569 CT C\n",
      "vdiff in left\n",
      "4 GTAAGTGCCCCG 7 CCCCCCCGGGCA\n",
      "chr19 1220729 G GC\n",
      "vdiff in left\n",
      "2 GAGACTCTGCCT 12 AAAAAAAAAAAA\n",
      "chr19 4097205 TA T\n",
      "vdiff in right\n",
      "2 GTTCCGTGGAGG 4 CCCTGCCCCTGC\n",
      "chr19 4110475 GCCCTGC G\n",
      "PASS\n",
      "3 GCCCTGCAGAGA 6 CCCCCCAGGGTA\n",
      "chr19 4117638 AC A\n",
      "vdiff in left\n",
      "3 GCCTGCTCCCAA 3 ACCCTCAGCAGG\n",
      "chr19 14093595 AACCCTCAGCAGGCTTAAGGAGGGGAGGCCCACCTTCCTCTGGTAG A\n",
      "PASS\n",
      "2 CAGCCTGTGCCT 6 GGGGGGCCCCCA\n",
      "chr19 50401797 TG T\n",
      "vdiff in right\n",
      "3 CCTTCAGTTTCT 6 GGGGGGCGTCTC\n",
      "chr19 50414747 TG T\n",
      "vdiff in right\n",
      "2 GCTGTTCGTGCT 2 GGAGCACAGCCT\n",
      "chr19 50416412 TG T\n",
      "vdiff in right\n",
      "4 TTTTCAGGCAGG 4 TTATTTTCTGCC\n",
      "chr22 41784278 GT G\n",
      "vdiff in left\n",
      "4 TTAGGAAAATTA 4 TCTTCCCCAAAT\n",
      "chrX 133026979 A ATCTTCCCCAAATTCACTTTTAGGCTTCCAAATTGTGTTGAGTTTTTCCATGTTATCTTTCAGTTGATCTAAACAGTGAGCTAAAAACTCATGAGCATCGTTCTGTGCATTGCCATGGAATATCTCTGCAGCTGCTGAAATGGCCTTTTTAAGATTCAAGAGTAACATCTCCTTGATTTCTATATTATAGGTATCTTTAAAAAAAAGTAGCCGTGCCAAGCACATGGTAAGAGCATTAAGGGGAATTTTACCCCATGGGAAACTCTGATTAAGTAAATCATCAGCAAACGATGGGATTGAAAGTAGAGACTGTAACACTGCATTCATATAACAGGTGTTTCCCAAATTGGGGAGGCCGTGGCATATTTTCTCTGGAAATAATTCAAAAAATAGTTTTAATTTATCCCACTTTGTGTAACCGTCACTATACCCTTGTTGTAACAGAAATACCAAAACCATTTTCTCAGTGAGAGCTTGGAGAAGACCAATGTCATCTAGGTAAGGATTTCCAGTGGCGTTCATGATGCATGAAGATTCACATTCCAATTTCTTATTCTCTTCTAACTCTTTTAACTTCAATTGTTTCTCTCGATTATAGCTTACACACCTCGAACAATCTGCCTTGGATTTCTTGTATTCTACAGAATTATTTTCTTTCAAGAATTCCTCATTCATCTCTGAGCTAGATGAGAGCATTCTTTTCCTCTTCTTGTGCTGATTTTCTGATAACTCTCCGCAAGTAAG\n",
      "PASS\n",
      "3 TCTTCTAATAAA 5 GAATAGTTTTTA\n",
      "chrY 12840645 AG A\n",
      "PASS\n",
      "3 TTACCCTCTTAG 9 TTTTTTTTTCTA\n",
      "chrY 12846292 G GT\n",
      "vdiff in right\n",
      "4 AAAAGAAAAGTA 7 TTTTTTTAATGT\n",
      "chrY 12856902 AT A\n",
      "vdiff in right\n",
      "1SCS014: Passcount is 12 and failcount is 49\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig 'chr1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] FILTER '-' is not defined in the header\n",
      "[W::vcf_parse] INFO 'donor' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse] INFO 'variation_type' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse] INFO 'subtype' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse_format] FORMAT 'filter' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse_format] FORMAT 'sample' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse_format] FORMAT 'AF' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse_format] FORMAT 'alt_depth' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse_format] FORMAT 'depth' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse_format] FORMAT 'no_calls' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse_format] FORMAT 'tissue' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse_format] FORMAT 'context' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse_format] FORMAT 'qual' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse_format] FORMAT 'age' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse] Contig 'chr2' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr3' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr4' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr5' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr6' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr7' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr8' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr9' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr11' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr12' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr13' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr14' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr15' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr16' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr17' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr19' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chr22' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrX' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'chrY' is not defined in the header. (Quick workaround: index the file with tabix.)\n"
     ]
    }
   ],
   "source": [
    "### Not touching 5\n",
    "\n",
    "def Substring(s):\n",
    "    ans, temp = 1, 1\n",
    "    # Traverse the string\n",
    "    for i in range(1, len(s)):\n",
    "        # If character is same as\n",
    "        # previous increment temp value\n",
    "        if (s[i] == s[i - 1]):\n",
    "            temp += 1\n",
    "        else:\n",
    "            ans = max(ans, temp)\n",
    "            temp = 1\n",
    " \n",
    "        ans = max(ans, temp)\n",
    "    # the required answer\n",
    "    return ans\n",
    "\n",
    "vcf = VCF(Atrust)\n",
    "\n",
    "failcount = 0\n",
    "passcount = 0 \n",
    "\n",
    "for v in vcf():\n",
    "    genome_seq_left = genome[v.CHROM][v.POS-12:v.POS]\n",
    "    genome_seq_right = genome[v.CHROM][v.POS:v.POS+12]\n",
    "    \n",
    "    valt = v.ALT[0]\n",
    "    vref = v.REF\n",
    "    if len(vref) > len(valt):\n",
    "        vdiff = vref.replace(valt, '', 1)\n",
    "    else: \n",
    "        vdiff = valt.replace(vref, '', 1)\n",
    "    \n",
    "    print(Substring(genome_seq_left), genome_seq_left, Substring(genome_seq_right), genome_seq_right)\n",
    "    if (vdiff * (Substring(genome_seq_left))) in genome_seq_left:\n",
    "        print( v.CHROM, v.POS, v.REF, v.ALT[0])            \n",
    "        failcount += 1\n",
    "        print('vdiff in left')\n",
    "    elif (vdiff * (Substring(genome_seq_right))) in genome_seq_right: \n",
    "        print( v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        failcount += 1\n",
    "        print('vdiff in right')\n",
    "    else:\n",
    "        print( v.CHROM, v.POS, v.REF, v.ALT[0])\n",
    "        passcount += 1\n",
    "        print('PASS')\n",
    "print(v.INFO.get('donor') + \":\", 'Passcount is', passcount, 'and failcount is', failcount) \n",
    "\n",
    " "
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
  },
  "kernelspec": {
   "display_name": "Python 3.8.9 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
